### ISSUE TITLE
Question in AdamGnT Optimizer Implementation

### ISSUE THREAD

**Issue by HowieRosier** [shibhansh/loss-of-plasticity#12]
Hi,

I've been analyzing the `AdamGnT` optimizer implementation in `lop/utils/AdamGnT.py`. I have a question about a specific line in the `step()` method, which seems to contain a potential logical bug.

The lines in question are:

```python
# ... (inside the step method)

bias_correction1 = 1 - beta1 ** state['step']
bias_correction = 1 - beta2 ** state['step']

# This line seems to have no effect on the 'bias_correction' variable
bias_correction.sqrt().div_(bias_correction1)

denom.div_(exp_avg)

# The final update uses the unmodified 'bias_correction'
p.data.addcdiv_(bias_correction, denom, value=-group['lr'])
```

My analysis suggests the following:

1.  The `.sqrt()` method in PyTorch is not an in-place operation. It returns a new tensor, which is then modified in-place by `.div_()`.
2.  Because the result of this chained operation is not reassigned, the original `bias_correction` variable (which holds `1 - beta2**t`) is never actually modified.
3.  As a result, the final update rule being executed is mathematically different from the standard Adam optimizer. The update appears to be:
    `p_new = p_old - lr * (m_t * (1 - beta2**t)) / (sqrt(v_t) + eps)`
    ...which is not the standard Adam update rule.

Could you please clarify if this is the intended behavior for the `AdamGnT` algorithm, or if this might be a typo? Perhaps the intention was to use an in-place `sqrt_()` and `div_()` or to reassign the result, like so:

```python
# Possible intended implementation
bias_correction = bias_correction.sqrt().div(bias_correction1)
```

Any clarification on this would be greatly appreciated. Thank you for your time and for the great work on this repository\!

**Comment by shibhansh** [comment]
Hey, 
You're right. That is a bug. I tested my implementation and found that this line has no effect: `bias_correction.sqrt().div_(bias_correction1)`

Maybe this is the reason why I never found AdamGnT to improve performance and why I did not use it in any experiments. Now I'm curious if it'll improve the performance of CBP when used with Adam. 

Why don't you send the correction you proposed as a pull request? Or I can just make the edit you recommended.
Thank you for catching the bug.

Cheers,
Shibhansh

**Comment by HowieRosier** [comment]
Yes, good idea. I'll try to send it. But before that, could you let me know what kinds of Adam you used in your experiments? Did you use the standard Adam optimizer?

**Comment by shibhansh** [comment]
Yes, I used standard Adam for all experiments that are reported in the paper. I also tested things with [AdamW](https://docs.pytorch.org/docs/stable/generated/torch.optim.AdamW.html), but there wasn't much of a difference between the performance of Adam and AdamW. You can send the pull request for either Adam or both Adam and AdamW.

**Comment by HowieRosier** [comment]
Hi Shibhansh,

I have revised some code in AdamGnT.py and sent it as a pull request. I have tested it to make sure it works the same way as standard Adam if no neurons are replaced but different to standard Adam if some neurons are replaced. Please also check it, and perhaps it can be used in CBP experiments. Thank you.

**Comment by shibhansh** [comment]
Hey, thank you for sending the pull request. I checked the changes, and AdamGnT works correctly now. Hopefully, I will get a chance to test it in CBP experiments soon.
